### PR TITLE
io: always forward flush to stdout and stderr

### PR DESCRIPTION
**Pull request by Darksonn**
Always call flush on the inner IO resource even if nothing has been written. This is necessary because someone could have written to stdout or stderr via some other object, and that data would go into the same shared buffer.

Closes: #4347

### CODE REVIEW

**Review by taiki-e — APPROVED** [review]
Could you add a test for #4347? Otherwise, LGTM.

**Comment by Darksonn** [comment]
I'm not sure how we can actually test it. We can assert that the write and flush succeeds, but we can't look at what actually got written.

**Comment by taiki-e** [comment]
Ah, it's indeed not easy to test it.

I think we can test this by adding the #4347's code to [tests-integration/src/bin](https://github.com/tokio-rs/tokio/tree/78e0f0b42a4f7a50f3986f576703e5a3cb473b79/tests-integration/src/bin) and doing something similar to the following shell script.

```sh
cargo run >tmp &
kill $!
grep <tmp "What's your name?: "
```

However, it is more complex and tricky than what I had imagined during the initial review, so I'm okay with merging this without adding a test.

**Review by taiki-e — APPROVED** [review]

**Comment by Darksonn** [comment]
It appears that the integration test keeps failing on Mac OS. Any ideas?

File: tests-integration/Cargo.toml [hunk 1]
```diff
@@ -8,6 +8,10 @@ publish = false
 [[bin]]
 name = "test-cat"
 
+[[bin]]
+name = "test-stdio-flush"
+required-features = ["rt-stdio"]
```
[the remark this answers is not in the document]
Comment (Darksonn) [reply tests-integration/Cargo.toml:13]: I didn't add it since the others weren't there, but lets try.

[the remark this answers is not in the document]
Comment (Darksonn) [reply tests-integration/Cargo.toml:13]: It seems like it still fails. :/

[the remark this answers is not in the document]
Comment (Darksonn) [reply tests-integration/Cargo.toml:13]: Ah.. I think I know why it fails. The `write` and the `flush` will both emit a `spawn_blocking` task, but there's nothing stopping them from being run out of order. So the flush could happen before the write, in which case the write is not flushed.

Thoughts? @taiki-e

**Pull request closed without merging**